service/handlers.go

```diff
@@ -838,6 +872,116 @@ func (h handler) listTargets(w http.ResponseWriter, r *http.Request) {
 	fmt.Fprint(w, string(data))
 }
 
+// Updates a target
+func (h handler) updateTarget(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	projectName := vars["projectName"]
+	targetName := vars["targetName"]
+
+	l := h.requestLogger(r, "op", "update-target", "project", projectName, "target", targetName)
+
+	level.Debug(l).Log("message", "validating authorization header for update target")
+	ah := r.Header.Get("Authorization")
+	a, err := credentials.NewAuthorization(ah)
+	if err != nil {
+		h.errorResponse(w, "error unauthorized, invalid authorization header", http.StatusUnauthorized)
+		return
+	}
+	if err := a.Validate(a.ValidateAuthorizedAdmin(h.env.AdminSecret)); err != nil {
+		h.errorResponse(w, "unauthorized", http.StatusUnauthorized)
+		return
+	}
+	level.Debug(l).Log("message", "reading request body")
+
+	var utr requests.UpdateTarget
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		level.Error(l).Log("message", "error reading request data", "error", err)
+		h.errorResponse(w, "error reading request data", http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.Unmarshal(reqBody, &utr); err != nil {
+		level.Error(l).Log("message", "error processing request", "error", err)
+		h.errorResponse(w, "error processing request", http.StatusBadRequest)
+		return
+	}
+
+	level.Debug(l).Log("message", "creating credential provider")
+	cp, err := h.newCredentialsProvider(*a, h.env, r.Header, credentials.NewVaultConfig, credentials.NewVaultSvc)
+	if err != nil {
+		level.Error(l).Log("message", "error creating credentials provider", "error", err)
+		h.errorResponse(w, "error creating credentials provider", http.StatusInternalServerError)
+		return
+	}
+
+	projectExists, err := cp.ProjectExists(projectName)
+	if err != nil {
+		level.Error(l).Log("message", "error determining if project exists", "error", err)
+		h.errorResponse(w, "error creating credentials provider", http.StatusInternalServerError)
+		return
+	}
+
+	if !projectExists {
+		level.Error(l).Log("message", "project does not exist")
+		h.errorResponse(w, "project does not exist", http.StatusNotFound)
+		return
+	}
+
+	targetExists, err := cp.TargetExists(projectName, targetName)
+	if err != nil {
+		level.Error(l).Log("message", "error retrieving target", "error", err)
+		h.errorResponse(w, "error retrieving target", http.StatusInternalServerError)
+		return
+	}
+	if !targetExists {
+		level.Error(l).Log("message", "target not found")
+		h.errorResponse(w, "target not found", http.StatusNotFound)
+		return
+	}
+
+	target, err := cp.GetTarget(projectName, targetName)
+	if err != nil {
+		level.Error(l).Log("message", "error retrieving existing target")
+		h.errorResponse(w, "error retrieving target", http.StatusInternalServerError)
+		return
+	}
+	targetType := target.Type
+
+	// merge request data into existing target struct for update data
+	if err := json.Unmarshal(reqBody, &target); err != nil {
```

Comment:
Since only the target properties are being updated, could the body be unmarshaled into a `TargetProperties` type then overwrite `target.Properties` with this type? I think that would do away with the code below.

https://github.com/argoproj-labs/argo-cloudops/pull/270/files#diff-ef779f0a29a099934f2007be0239b2ece9a9614de96dafcb64b7d18f4ebd4b83R957-R959

Reply:
If we do that it would overwrite the target property values you don't provide with blank values.  So if we did that, you would have to provide all of the existing values in the PATCH request even the ones that aren't being updated.